changefeedccl: filter out virtual computed columns from events

sherman-grewal opened the pull request:
changefeedccl: filter out virtual computed columns from events

Resolves #74688

Release note (backward-incompatible change): Changefeeds will
now filter out virtual computed columns from events by default.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74916)
<!-- Reviewable:end -->

sherman-grewal commented on pkg/ccl/changefeedccl/encoder.go:
```diff
@@ -199,8 +199,11 @@ func (e *jsonEncoder) EncodeValue(_ context.Context, row encodeRow) ([]byte, err
 	var after map[string]interface{}
 	if !row.deleted {
 		columns := row.tableDesc.PublicColumns()
```
I considered adding a method to the TableDescriptor that only stores non-virtual columns, but I believe that this would require us to extend our column cache to store non-virtual columns. I am not sure on how expensive this would be, but I would love to get some thoughts on whether or not we should go in that direction (as apposed to filtering out virtual columns that are returned from `PublicColumns()`).

shermanCRL commented:
Nice. What was the previous behavior, nil? If so, @HonoreDB should this be noted as a breaking change?

sherman-grewal commented:
> Nice. What was the previous behavior, nil? If so, @HonoreDB should this be noted as a breaking change?

Yes, virtual computed columns would show up as nil in the events. It is a breaking change, but I was not sure on how to indicate that.

shermanCRL commented:
It can be added to the release notes above: https://wiki.crdb.io/wiki/spaces/CRDB/pages/186548364/Release+notes#Backward-incompatible-change

HonoreDB reviewed (changes requested):
Because this is a breaking change, we will need an option to include the legacy behavior. Thinking about the ideas we discussed last week I think the best one is `WITH virtual_columns=null`, so that we leave room for explicitly specifying the new default behavior with `=omitted` and eventually specifying `=computed` once we support that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74916#-:-MthiTje1PFEMouYbaGE:b-jgxcyh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @sherman-grewal)

___
*[pkg/ccl/changefeedccl/encoder.go, line 201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74916#-Mtdo4c144SqE4Fqs2xW-r1-201:-MthjvgVBIyGlYucp4T6:b-4coqxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d6e6a9d9aee302d034ff577dd9a323208cb577/pkg/ccl/changefeedccl/encoder.go#L201)):*
<details><summary><i>Previously, sherman-grewal wrote…</i></summary><blockquote>

I considered adding a method to the TableDescriptor that only stores non-virtual columns, but I believe that this would require us to extend our column cache to only store non-virtual columns. I am not sure on how expensive this would be, but I would love to get some thoughts on whether or not we should go in that direction (as apposed to filtering out virtual columns that are returned from `PublicColumns()`).
</blockquote></details>

Your current approach seems good to me.
___
*[pkg/ccl/changefeedccl/changefeedbase/validate.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74916#-MthjY2h0VKJHWRKGMKF:-MthjY2h0VKJHWRKGMKG:b-qqajea) ([raw file](https://github.com/cockroachdb/cockroach/blob/79bb0837e3b9cd028f0ae013624db7840548632d/pkg/ccl/changefeedccl/changefeedbase/validate.go#L64)):*
> ```Go
> 		if col.IsVirtual() {
> 			warnings = append(warnings,
> 				errors.Errorf("Changefeeds will emit null values for virtual column %s in table %s", col.ColName(), tableDesc.GetName()),
> ```

We still want this warning, the phrasing just needs to change.
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 2304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74916#-Mthjk0Y1VTzd14gkrGi:-Mthjk0Y1VTzd14gkrGj:b-4m3hag) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d6e6a9d9aee302d034ff577dd9a323208cb577/pkg/ccl/changefeedccl/changefeed_test.go#L2304)):*
> ```Go
> 		sqlDB.Exec(t, `UPDATE cc SET b=10 WHERE a=1`)
> 		assertPayloads(t, cc, []string{
> 			`cc: [1]->{"after": {"a": 1, "b": 10}, "before": {"a": 1, "b": 1}}`,
> ```

Please also add a test with `format=avro`.


<!-- Sent from Reviewable.io -->

sherman-grewal reviewed (commented):
> Reviewed 4 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74916)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB)

___
*[pkg/ccl/changefeedccl/changefeedbase/validate.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74916#-MthjY2h0VKJHWRKGMKF:-MtiqUZ25_KKCMlIaQH8:bu9gagv) ([raw file](https://github.com/cockroachdb/cockroach/blob/79bb0837e3b9cd028f0ae013624db7840548632d/pkg/ccl/changefeedccl/changefeedbase/validate.go#L64)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

We still want this warning, the phrasing just needs to change.
</blockquote></details>

Done. Thanks for catching this!
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 2304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74916#-Mthjk0Y1VTzd14gkrGi:-MtiqVT15POcmRrNaxcJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d6e6a9d9aee302d034ff577dd9a323208cb577/pkg/ccl/changefeedccl/changefeed_test.go#L2304)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Please also add a test with `format=avro`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
changefeedccl: filter out virtual computed columns from events

Release note (backward-incompatible change): Changefeeds will
now filter out virtual computed columns from events by default.

HonoreDB reviewed (approved):
Very nicely done! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74916#-:-MtnUVlJ819w7UABzeqJ:bc6j2f6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @HonoreDB and @sherman-grewal)

___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 2288 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74916#-MtnT8r7AFm-V4SIagbs:-MtnT8r7AFm-V4SIagbt:b-83sjo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e3d60c8b3aa7dcae11526c52969767e4e3e244e/pkg/ccl/changefeedccl/changefeed_test.go#L2288)):*
> ```Go
> 	defer log.Scope(t).Close(t)
> 
> 	tests := map[string]struct {
> ```

Nice.


<!-- Sent from Reviewable.io -->

sherman-grewal commented:
bors r+
